Date: Fri Mar  4 22:05:24 CET 2005
From: Vojtech Pavlik <vojtech@suse.cz>
Subject: Fix ALPS DualPoint stick

=====================================================================================

ChangeSet@1.2068.1.2, 2005-03-04 20:19:05+01:00, vojtech@suse.cz
  input: Fix inverted conditions in ALPS DualPoint stick packet
         decoding.
  Signed-off-by: Vojtech Pavlik <vojtech@suse.cz>

ChangeSet@1.2072, 2005-03-04 21:55:23+01:00, vojtech@suse.cz
  input: Fix ALPS DualPoint stick buttons. Testing shows that they're
         at the same place as the pad ones, but packet has z==127.
  Signed-off-by: Vojtech Pavlik <vojtech@suse.cz>

=====================================================================================
  
 alps.c |   26 +++++++++---------------
 1 files changed, 10 insertions(+), 16 deletions(-)

=====================================================================================

diff -Nru a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
--- a/drivers/input/mouse/alps.c	2005-03-04 22:00:33 +01:00
+++ b/drivers/input/mouse/alps.c	2005-03-04 22:00:33 +01:00
@@ -124,8 +124,8 @@
 
 		/* Relative movement packet */
  		if (z == 127) {
-			input_report_rel(dev2, REL_X,  (x > 383 ? x : (x - 768)));
-			input_report_rel(dev2, REL_Y, -(y > 255 ? y : (x - 512)));
+			input_report_rel(dev2, REL_X,  (x > 383 ? (x - 768) : x));
+			input_report_rel(dev2, REL_Y, -(y > 255 ? (x - 512) : y));
 			input_sync(dev2);
 			return;
 		}
diff -Nru a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
--- a/drivers/input/mouse/alps.c	2005-03-04 22:00:49 +01:00
+++ b/drivers/input/mouse/alps.c	2005-03-04 22:00:49 +01:00
@@ -115,20 +115,14 @@
 	ges = packet[2] & 1;
 	fin = packet[2] & 2;
 
-	/* Dualpoint has stick buttons in byte 0 */
-	if (priv->i->flags & ALPS_DUALPOINT) {
-	
-		input_report_key(dev2, BTN_LEFT,    packet[0]       & 1);    
-		input_report_key(dev2, BTN_MIDDLE, (packet[0] >> 2) & 1);
-		input_report_key(dev2, BTN_RIGHT,  (packet[0] >> 1) & 1);
-
-		/* Relative movement packet */
- 		if (z == 127) {
-			input_report_rel(dev2, REL_X,  (x > 383 ? (x - 768) : x));
-			input_report_rel(dev2, REL_Y, -(y > 255 ? (x - 512) : y));
-			input_sync(dev2);
-			return;
-		}
+	if ((priv->i->flags & ALPS_DUALPOINT) && z == 127) {
+		input_report_key(dev2, BTN_LEFT,   left);    
+		input_report_key(dev2, BTN_RIGHT,  right);
+		input_report_key(dev2, BTN_MIDDLE, middle);
+		input_report_rel(dev2, REL_X,  (x > 383 ? (x - 768) : x));
+		input_report_rel(dev2, REL_Y, -(y > 255 ? (x - 512) : y));
+		input_sync(dev2);
+		return;
 	}
 
 	/* Convert hardware tap to a reasonable Z value */
